test/models/content_view_component_test.rb

```diff
@@ -229,25 +229,29 @@ def test_needs_publish
                                                       :action => "publish")
       assert_equal @composite.needs_publish?, true
       @composite.create_new_version
+      @composite.reload
```

Comment:
Can `create_new_version` somehow invalidate the `latest_version_object` cache on the object?

Reply:
- Rails load the `@composite` content view. During this, it already loads the latest_version_object relation "has_one -> latest"
- create_new_version creates a new version but the reference in the already loaded `@composite` is not updated.

Reply:
Yes, I got that. That's why I was asking if you can somehow flush that cached `latest_version_object` from Rails. Because I'm worried you might end up serving some outdated data somewhere. Looks like this is a Rails 7.1 feature: https://www.shakacode.com/blog/rails-7-1-allows-resetting-singular-associations/

So I think you can some something like:
```ruby
# Rails 7.1+
reset_latest_version_object if respond_to?(:reset_latest_version_object)
```

Reply:
I seem to be on Rails 6.1.7.6. (I didn't think we had upgraded to 7.1 already..)

Anyway, `reload` seems fine to me for the time being.

Reply:
> I seem to be on Rails 6.1.7.6. (I didn't think we had upgraded to 7.1 already..)

Yes, we're not there yet. I wish we were, but not yet. We're working on it.

> Anyway, `reload` seems fine to me for the time being.

My point was that this can also break things at runtime. Previously you could count on `latest_version_object` being up to date, but after this it may be stale. For example, if a controller creates the new object and then uses `latest_version_object` it will show the old data.

I don't know if that happens and if we have any coverage for it, but it's an API change that can be breaking.

Reply:
run a `reload` inside create_new_version?

Reply:
There is a performance overhead with `reload`, because it actually loads an object which is an additional DB query that it doesn't need in some cases (which Rails 7.1 solves with a new method).

At this point I'm a bit uncertain, but it's a risk. When I take a look where it's called, it's probably a small risk.

Reply:
Is it really necessary to reload it? AFAIK, if API / UI  is used, you run a action, like "Create a new Version" and then you run another action - which would be a completely new request and therefore all the structure will be reloaded automatically. I think, we are trying to solve a "unit-test-issue" which doesn't occur in production.

Reply:
No, I'm not saying you need to call `reload`. All I'm saying is that there is a cache invalidation that doesn't happen and that's a risk.

For example, if you have a REST API that is something like `POST /content_views/:id/versions` (so effectively an API controller) and the implementation is roughly:
```ruby
@content_view.create_new_version
{
  'id' => @content_view.id,
  'latest' => @content_view.latest_version_object,
}
```

Before this patch it would work. After this patch it would show incorrect data.

It would be very easy to do this, because the actual use is in RABL. Also note that there are other indirections, such as `latest_version` and `latest_version_id`. So these do show up here:

https://github.com/Katello/katello/blob/3dcae86ca41974534ec67771364ed6b312351e92/app/views/katello/api/v2/content_views/base.json.rabl#L8

So all I'm saying is that we must verify such code doesn't exist, or if it does, deal with it. If it does exist, there is a very good chance there are no tests for such a thing, so it doesn't show up as a failure. It's very easy to introduce a regression here.

Reply:
So, should we:
1. do a reload in create_new_version or
2. run a reload after create_new_version in the test?

Reply:
I think that from reading the source of the actions it's probably safe to assume it's not a problem, but I'd still leave at least a TODO comment for Rails 7.1 that it can safely invalidate the cached entry.

Reply:
ok